source/common/ur_params.hpp

```diff
@@ -0,0 +1,7780 @@
+}
+
+inline std::ostream &
+operator<<(std::ostream &os, const struct ur_device_retain_params_t *params) {
+
+    os << ".hDevice = ";
+
+    os << *(params->phDevice);
+
+    return os;
+}
+
+inline std::ostream &
+operator<<(std::ostream &os, const struct ur_device_release_params_t *params) {
+
+    os << ".hDevice = ";
+
+    os << *(params->phDevice);
+
+    return os;
+}
+
+inline std::ostream &
+operator<<(std::ostream &os,
+           const struct ur_device_partition_params_t *params) {
+
+    os << ".hDevice = ";
+
+    os << *(params->phDevice);
+
+    os << ", ";
+    os << ".pProperties = ";
+
+    serializePtr(os, *(params->ppProperties));
+
+    os << ", ";
+    os << ".NumDevices = ";
+
+    os << *(params->pNumDevices);
+
+    os << ", ";
+    os << ".phSubDevices = [";
+    for (size_t i = 0;
+         *(params->pphSubDevices) != NULL && i < *params->pNumDevices; ++i) {
+        if (i != 0) {
+            os << ", ";
+        }
+
+        os << (*(params->pphSubDevices))[i];
+    }
+    os << "]";
+
+    os << ", ";
+    os << ".pNumDevicesRet = ";
+
+    serializePtr(os, *(params->ppNumDevicesRet));
+
+    return os;
+}
+
+inline std::ostream &
+operator<<(std::ostream &os,
+           const struct ur_device_select_binary_params_t *params) {
+
+    os << ".hDevice = ";
+
+    os << *(params->phDevice);
+
+    os << ", ";
+    os << ".pBinaries = ";
+
+    serializePtr(os, *(params->ppBinaries));
+
+    os << ", ";
+    os << ".NumBinaries = ";
+
+    os << *(params->pNumBinaries);
+
+    os << ", ";
+    os << ".pSelectedBinary = ";
+
+    serializePtr(os, *(params->ppSelectedBinary));
+
+    return os;
+}
+
+inline std::ostream &
+operator<<(std::ostream &os,
+           const struct ur_device_get_native_handle_params_t *params) {
+
+    os << ".hDevice = ";
+
+    os << *(params->phDevice);
+
+    os << ", ";
+    os << ".phNativeDevice = ";
+
+    serializePtr(os, *(params->pphNativeDevice));
+
+    return os;
+}
+
+inline std::ostream &
+operator<<(std::ostream &os,
+           const struct ur_device_create_with_native_handle_params_t *params) {
+
+    os << ".hNativeDevice = ";
+
+    os << *(params->phNativeDevice);
+
+    os << ", ";
+    os << ".hPlatform = ";
+
+    os << *(params->phPlatform);
+
+    os << ", ";
+    os << ".phDevice = ";
+
+    serializePtr(os, *(params->pphDevice));
+
+    return os;
+}
+
+inline std::ostream &
+operator<<(std::ostream &os,
+           const struct ur_device_get_global_timestamps_params_t *params) {
+
+    os << ".hDevice = ";
+
+    os << *(params->phDevice);
+
+    os << ", ";
+    os << ".pDeviceTimestamp = ";
+
+    serializePtr(os, *(params->ppDeviceTimestamp));
+
+    os << ", ";
+    os << ".pHostTimestamp = ";
+
+    serializePtr(os, *(params->ppHostTimestamp));
+
+    return os;
+}
+
+template <typename T> inline void serializePtr(std::ostream &os, T *ptr) {
+    if (ptr == nullptr) {
+        os << "nullptr";
+    } else if constexpr (std::is_same_v<T, void> ||
```

Comment:
What if it's volatile ptr? Probably it never is, but it's safer to use https://en.cppreference.com/w/cpp/types/remove_cv and then compare. Or you might use https://en.cppreference.com/w/cpp/types/decay if you want to support arrays as well.